File: ios/HybridThumbnail.swift
import AVFoundation
import Foundation
import NitroModules

final class HybridThumbnail: HybridThumbnailSpec {
  func create(
    options: NativeThumbnailOptions
  ) throws -> Promise<NativeThumbnailResult> {
    return Promise.async {
      let (asset, isRemote) = try Self.makeAsset(options.url, headers: options.headers)

      let gen = AVAssetImageGenerator(asset: asset)
      gen.appliesPreferredTrackTransform = true
      gen.maximumSize = CGSize(width: options.maxWidth, height: options.maxHeight)
      let tol = CMTime(value: CMTimeValue(options.timeToleranceMs), timescale: 1000)
      gen.requestedTimeToleranceBefore = tol
      gen.requestedTimeToleranceAfter = tol

      let time = CMTime(value: CMTimeValue(options.timeStamp), timescale: 1000)
      let cg: CGImage
      do {
        cg = try gen.copyCGImage(at: time, actualTime: nil)
      } catch {
        let ns = error as NSError
        if isRemote && ns.domain == NSURLErrorDomain {
          throw Self.err(
            "REMOTE_FETCH_FAILED",
            "Could not fetch remote video: \(error.localizedDescription)")
        }
        throw Self.err(
          "DECODE_FAILED",
          "Could not extract frame: \(error.localizedDescription)")
      }

      guard
        let data = ThumbnailEncoder.encode(
          cg, format: options.format, quality: options.quality)
      else {
        throw Self.err("UNSUPPORTED_FORMAT", "Cannot encode as \(options.format)")
      }

      let outURL = try Self.outputURL(
        format: options.format, cacheName: options.cacheName)
      do {
        try data.write(to: outURL, options: .atomic)
      } catch {
        throw Self.err("WRITE_FAILED", error.localizedDescription)
      }

      return NativeThumbnailResult(
        path: outURL.absoluteString,
        size: Double(data.count),
        mime: options.format == "png" ? "image/png" : "image/jpeg",
        width: Double(cg.width),
        height: Double(cg.height))
    }
  }

  // MARK: helpers

  /// Nitro surfaces only the error *message* string to JS, so encode the typed
  /// code as a leading `[CODE] message` prefix that `toThumbnailError` parses.
  private static func err(_ code: String, _ message: String) -> RuntimeError {
    return RuntimeError("[\(code)] \(message)")
  }

  /// Build an AVURLAsset for a local file path or a remote http(s) URL (with headers).
  /// Returns the asset and whether it is remote (for error-code mapping).
  private static func makeAsset(
    _ raw: String, headers: [String: String]?
  ) throws -> (AVURLAsset, Bool) {
    if raw.hasPrefix("http://") || raw.hasPrefix("https://") {
      guard let u = URL(string: raw) else {
        throw err("INVALID_URL", "Malformed URL: \(raw)")
      }
      var options: [String: Any] = [:]
      if let headers = headers, !headers.isEmpty {
        options["AVURLAssetHTTPHeaderFieldsKey"] = headers
      }
      return (AVURLAsset(url: u, options: options), true)
    }
    if raw.hasPrefix("file://"), let u = URL(string: raw) {
      guard FileManager.default.fileExists(atPath: u.path) else {
        throw err("FILE_NOT_FOUND", "No file at \(u.path)")
      }
      return (AVURLAsset(url: u), false)
    }
    if raw.hasPrefix("/") {
      guard FileManager.default.fileExists(atPath: raw) else {
        throw err("FILE_NOT_FOUND", "No file at \(raw)")
      }
      return (AVURLAsset(url: URL(fileURLWithPath: raw)), false)
    }
    throw err("INVALID_URL", "Unsupported URL: \(raw)")
  }

  private static func outputURL(format: String, cacheName: String?) throws -> URL {
    let dir = FileManager.default
      .urls(for: .cachesDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("thumbnails", isDirectory: true)
    try FileManager.default.createDirectory(
      at: dir, withIntermediateDirectories: true)
    let ext = format == "png" ? "png" : "jpg"
    let base =
      cacheName?.isEmpty == false ? cacheName! : "thumb-\(UUID().uuidString)"
    return dir.appendingPathComponent("\(base).\(ext)")
  }
}

File: android/src/main/java/com/margelo/nitro/nitrothumbnail/HybridThumbnail.kt
package com.margelo.nitro.nitrothumbnail

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.io.File
import java.util.UUID

@DoNotStrip
class HybridThumbnail : HybridThumbnailSpec() {
  override fun create(options: NativeThumbnailOptions): Promise<NativeThumbnailResult> {
    return Promise.async {
      // cacheName dedup: if a thumbnail with this name already exists, return it.
      if (!options.cacheName.isNullOrEmpty()) {
        val candidate = outputFile(options.format, options.cacheName)
        existingResult(candidate, options.format)?.let { return@async it }
      }

      val retriever = MediaMetadataRetriever()
      try {
        setSource(retriever, options.url, options.headers)

        val maxW = options.maxWidth.toInt()
        val maxH = options.maxHeight.toInt()
        val timeUs = (options.timeStamp * 1000L).toLong() // ms -> microseconds
        val sync = if (options.onlySyncedFrames)
          MediaMetadataRetriever.OPTION_CLOSEST_SYNC
        else
          MediaMetadataRetriever.OPTION_CLOSEST

        val frame: Bitmap = extractFrame(retriever, timeUs, sync, maxW, maxH)
          ?: throw err("DECODE_FAILED", "Could not extract a frame at ${options.timeStamp}ms")

        val bytes = ThumbnailEncoderKt.encode(frame, options.format, options.quality)
          ?: throw err("UNSUPPORTED_FORMAT", "Cannot encode as ${options.format}")

        val out = outputFile(options.format, options.cacheName)
        try {
          out.writeBytes(bytes)
        } catch (e: Exception) {
          throw err("WRITE_FAILED", e.message ?: "write failed")
        }

        enforceLimit(out.parentFile, (options.dirSize * 1024 * 1024).toLong())

        NativeThumbnailResult(
          path = Uri.fromFile(out).toString(),
          size = bytes.size.toDouble(),
          mime = ThumbnailEncoderKt.mimeFor(options.format),
          width = frame.width.toDouble(),
          height = frame.height.toDouble(),
        )
      } finally {
        retriever.release()
      }
    }
  }

  // MARK: helpers

  /** Nitro surfaces only the error message to JS, so encode the code as a "[CODE] message" prefix. */
  private fun err(code: String, message: String) = RuntimeException("[$code] $message")

  /** Point the retriever at a local file or a remote http(s) URL (with headers). */
  private fun setSource(
    retriever: MediaMetadataRetriever,
    raw: String,
    headers: Map<String, String>?,
  ) {
    when {
      raw.startsWith("http://") || raw.startsWith("https://") -> {
        try {
          retriever.setDataSource(raw, headers ?: emptyMap())
        } catch (e: Exception) {
          throw err("REMOTE_FETCH_FAILED", "Could not fetch remote video: ${e.message}")
        }
      }
      else -> {
        val path = when {
          raw.startsWith("file://") -> Uri.parse(raw).path ?: raw.removePrefix("file://")
          raw.startsWith("/") -> raw
          else -> throw err("INVALID_URL", "Unsupported URL: $raw")
        }
        val file = File(path)
        if (!file.exists()) throw err("FILE_NOT_FOUND", "No file at $path")
        try {
          retriever.setDataSource(file.absolutePath)
        } catch (e: Exception) {
          throw err("DECODE_FAILED", "Could not open video: ${e.message}")
        }
      }
    }
  }

  private fun extractFrame(
    retriever: MediaMetadataRetriever,
    timeUs: Long,
    option: Int,
    maxW: Int,
    maxH: Int,
  ): Bitmap? {
    if (Build.VERSION.SDK_INT >= 27) {
      retriever.getScaledFrameAtTime(timeUs, option, maxW, maxH)?.let { return it }
    }
    // Pre-27 (or null result): full frame then scale down preserving aspect.
    val full = retriever.getFrameAtTime(timeUs, option) ?: return null
    val (w, h) = ThumbnailEncoderKt.targetSize(full.width, full.height, maxW, maxH)
    if (w == full.width && h == full.height) return full
    val scaled = Bitmap.createScaledBitmap(full, w.coerceAtLeast(1), h.coerceAtLeast(1), true)
    if (scaled != full) full.recycle()
    return scaled
  }

  private fun outputFile(format: String, cacheName: String?): File {
    val ctx = NitroModules.applicationContext
      ?: throw err("WRITE_FAILED", "No Android context available")
    val dir = File(ctx.cacheDir, "thumbnails").apply { mkdirs() }
    val ext = if (format == "png") "png" else "jpg"
    val base = if (!cacheName.isNullOrEmpty()) cacheName else "thumb-${UUID.randomUUID()}"
    return File(dir, "$base.$ext")
  }

  /** Read an existing thumbnail's metadata (for cacheName dedup), or null. */
  private fun existingResult(file: File, format: String): NativeThumbnailResult? {
    if (!file.exists()) return null
    val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.absolutePath, opts)
    if (opts.outWidth <= 0 || opts.outHeight <= 0) return null
    return NativeThumbnailResult(
      path = Uri.fromFile(file).toString(),
      size = file.length().toDouble(),
      mime = ThumbnailEncoderKt.mimeFor(format),
      width = opts.outWidth.toDouble(),
      height = opts.outHeight.toDouble(),
    )
  }

  /** Enforce the dirSize cap via LRU eviction of the thumbnails directory. */
  private fun enforceLimit(dir: File?, capBytes: Long) {
    if (dir == null || capBytes <= 0) return
    val files = dir.listFiles() ?: return
    val entries = files.filter { it.isFile }
      .map { Triple(it.absolutePath, it.length(), it.lastModified()) }
    for (path in ThumbnailEncoderKt.filesToEvict(entries, capBytes)) {
      File(path).delete()
    }
  }
}
